XenAdmin/Dialogs/InstallCertificateDialog.cs

```diff
@@ -0,0 +1,336 @@
+        }
+
+        private void buttonAddCertificate_Click(object sender, EventArgs e)
+        {
+            using (var dlg = new OpenFileDialog
+            {
+                Title = Messages.CERTIFICATE_SELECT_CERTIFICATE_TITLE,
+                Multiselect = true,
+                CheckFileExists = true,
+                CheckPathExists = true,
+                Filter = Messages.CERTIFICATE_FILETYPES
+            })
+            {
+                if (dlg.ShowDialog(this) == DialogResult.OK)
+                {
+                    var rows = new List<DataGridViewRow>();
+
+                    foreach (var name in dlg.FileNames)
+                    {
+                        var row = new CertificateRow(name);
+
+                        if (dataGridViewCertificates.Rows.Cast<CertificateRow>().All(r => r.FileName != row.FileName))
+                            rows.Add(row);
+                        //All() returns true for empty collection which is correct for this predicate
+                    }
+
+                    dataGridViewCertificates.Rows.AddRange(rows.ToArray());
+                    HideAllErrors();
+                    UpdateButtons();
+                }
+            }
+        }
+
+        private void buttonRemove_Click(object sender, EventArgs e)
+        {
+            if (dataGridViewCertificates.SelectedRows.Count != 1)
+                return;
+
+            var row = dataGridViewCertificates.SelectedRows[0];
+            dataGridViewCertificates.Rows.Remove(row);
+
+            HideAllErrors();
+            UpdateButtons();
+        }
+
+        private void buttonCancel_Click(object sender, EventArgs e)
+        {
+            Close();
+        }
+
+        private void buttonInstall_Click(object sender, EventArgs e)
+        {
+            var certificateFiles = new List<string>();
+            foreach (var gridRow in dataGridViewCertificates.Rows)
+            {
+                if (gridRow is CertificateRow row)
+                    certificateFiles.Add(row.FileName);
+            }
+
+            _action = new InstallCertificateAction(_host, textBoxKey.Text.Trim(), textBoxCertificate.Text.Trim(),
+                certificateFiles, DateConverter);
+
+            _action.RequestReconnection += _action_RequestReconnection;
+            _action.Changed += _action_Changed;
+            _action.Completed += _action_Completed;
+
+            labelKeyError.Visible = false;
+            labelCertificateError.Visible = false;
+            labelChainError.Visible = false;
+
+            UpdateProgress();
+            pictureBox1.Visible = false;
+            progressBar1.Visible = true;
+            tlpActionProgress.Visible = true;
+            
+            _action.RunAsync();
+            UpdateButtons();
+        }
+
+        private void _action_RequestReconnection(IXenConnection conn)
+        {
+            Program.Invoke(Program.MainWindow, () => XenConnectionUI.BeginConnect(conn, false, null, false));

```

Comment:
Shouldn't we only try to reconnect if the connection is disconnected?

Reply:
I didn't bother because the connection's BeginConnect does check internally (it reconnects only if the connection task is null).